tools/query_verification/src/main/java/com/google/bigquery/QueryVerifier.java

```diff
@@ -112,5 +117,43 @@ public void verifyDataAware() throws InterruptedException {
         System.out.println("Data-Aware Verification Completed");
     }
 
+    /**
+     * Finds extra and missing results by locating the differences between the results.
+     * @param migratedResults Parsed results returned from BQ
+     * @param originalResults Parsed results returned from original data warehouse service
+     * @return Differences classified as either extra or missing from migrated results.
+     */
+    public ResultDifferences compareResults(List<QueryJobResults> migratedResults, List<QueryJobResults> originalResults) {
+        // Check if same amount of queries were run
+        if (migratedResults.size() == originalResults.size()) {
+            // Rows present in migrated query results, but not original query results
+            List<List<Object>> extraResults = new ArrayList<List<Object>>();
+
+            // Rows present in original query results, but not migrated query results
+            List<List<Object>> missingResults = new ArrayList<List<Object>>();
+
+            for (int i = 0; i < migratedResults.size(); i++) {
+                QueryJobResults migratedJobResults = migratedResults.get(i);
+
+                QueryJobResults originalJobResults = originalResults.get(i);
+                Set<List<Object>> missingResultsSet = originalJobResults.results();
+
+                for (List<Object> migratedQueryResults : migratedJobResults.results()) {
+                    // Rows that exist in both results are removed from missing results set
+                    if (!missingResultsSet.remove(migratedQueryResults)) {
+                        // Rows in the migrated results that don't exist in original results are classified as extra in migrated results
+                        extraResults.add(migratedQueryResults);
+                    }
+                }
+
+                // Any leftover rows in the original results without a match are classified as missing in migrated results
+                missingResults.addAll(missingResultsSet);
+            }
+
+            return ResultDifferences.create(extraResults, missingResults);
+        } else {
+            return null;
```

Comment:
return null may cause unexpected issue if other parts didn't handle carefully. Consider using Optional<ResultDifferences> as return type

Reply:
Seconding this. I also wonder why we even need this else branch since I think we should be able to produce a result difference no matter what the size of each result set is.

Reply:
Updated. The condition was just checking if the same number of query statements were run. If they weren't, it meant one of the parameters had results from a query that wasn't ran in the other